ext/openssl/ossl_ssl.c

```diff
@@ -224,6 +225,199 @@ ossl_ssl_verify_callback(int preverify_ok, X509_STORE_CTX *ctx)
+  return rb_funcall(args->proc, rb_intern("call"), 6,
+                    args->write_p,
+                    args->version,
+                    args->content_type,
+                    args->details1,
+                    args->details2,
+                    args->buf);
+}
+
+static void imp_ssl_msg_callback(int write_p, int version, int content_type,
+                                        const void *buf, size_t len, SSL *ssl, void *arg)
+{
+  struct ossl_msg_cb_args args = { .content_type = Qnil,
+                                   .details1 = Qnil,
+                                   .details2 = Qnil,
+                                   .buf = Qnil };
+  const char* bp = buf;
+
+  args.proc = (VALUE)arg;
+  args.write_p = write_p ? Qtrue : Qfalse;
+
+  if (version == SSL3_VERSION ||
+        version == TLS1_VERSION ||
+        version == TLS1_1_VERSION ||
+        version == TLS1_2_VERSION ||
+        version == TLS1_3_VERSION ||
+        version == DTLS1_VERSION || version == DTLS1_BAD_VER) {
+    const char *str_version = lookup(version, ssl_versions, "???");
+
+    args.version = rb_str_new_cstr(str_version);
+    switch (content_type) {
+      case SSL3_RT_CHANGE_CIPHER_SPEC:
+          /* type 20 */
+          args.content_type = rb_str_new_cstr("ChangeCipherSpec");
+          break;
+      case SSL3_RT_ALERT:
+          /* type 21 */
+          args.content_type = rb_str_new_cstr("Alert");
+          if (len == 2) {
+              switch (bp[0]) {
+              case 1:
+                  args.details1 = rb_str_new_cstr("warning");
+                  break;
+              case 2:
+                  args.details1 = rb_str_new_cstr("fatal");
+                  break;
+              }
+              args.details2 = rb_str_new_cstr(lookup((int)bp[1], alert_types, "???"));
+          } else {
+            args.details1 = rb_str_new_cstr("???");
+          }
+          break;
+      case SSL3_RT_HANDSHAKE:
+          /* type 22 */
+          args.content_type = rb_str_new_cstr("Handshake");
+          if (len > 0) {
+              args.details1 = rb_str_new_cstr(lookup((int)bp[0], handshakes, "???"));
+          } else {
+            args.details1 = rb_str_new_cstr("???");
+          }
+          break;
+      case SSL3_RT_APPLICATION_DATA:
+          /* type 23 */
+          args.content_type = rb_str_new_cstr("ApplicationData");
+          break;
+      case SSL3_RT_HEADER:
+          /* type 256 */
+          args.content_type = rb_str_new_cstr("RecordHeader");
+          break;
+      case SSL3_RT_INNER_CONTENT_TYPE:
+          /* type 257 */
+          args.content_type = rb_str_new_cstr("InnerContent");
+          break;
+      default:
+          args.content_type = rb_sprintf("Unknown (content_type=%d)", content_type);
+      }
+  } else {
+    args.version = rb_sprintf("Not TLS data or unknown version (version=%d, content_type=%d)", version, content_type);
+  }
+
+  args.buf = rb_str_new(bp, len);
```

Comment:
Any Ruby object allocation that isn't an immediate value can potentially fail, so these `rb_str_new*()` calls need to be moved to inside `rb_protect()`.